L2-800: Dropdown Select Account

lmuntaner opened the pull request:
# Motivation

User sees a nice Dropdown.

# Changes

* New Dropdown UI component.
* SelectAccountDropdown uses new Dropdown UI Component.

# Tests

* Test for Dropdown UI component
* Test for SelectAccountDropdown component

lmuntaner commented:
@peterpeterparker @mstrasinskis please review

lmuntaner commented:
Dropdown is still a native dropdown.

New UX:

https://user-images.githubusercontent.com/4550653/177561396-79f9b402-8a33-4f31-8b34-945b5ec627ad.mov

Screenshots:

![Screenshot 2022-07-06 at 14 30 49](https://user-images.githubusercontent.com/4550653/177561454-d7a234c2-ade1-450c-9021-6688c8604c4c.jpg)
![Screenshot 2022-07-06 at 14 31 07](https://user-images.githubusercontent.com/4550653/177561456-db052aaa-6724-4cd3-bf5e-dbce11447a3c.jpg)

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Dropdown.svelte:
```diff
@@ -0,0 +1,52 @@
+  import IconExpandMore from "../../icons/IconExpandMore.svelte";
+  import type { SelectOption } from "../../types/common";
+
+  // Do not allow to bind objects, since they are mutable.
```
what do you mean?

lmuntaner replied:
I first set the value of the options to be the Account object, but then, when the update call finished and the store was substituted with the new values, the "selected" account disappeared. The reason was that the old value was the account object got during the query, and the object even though it had the same information it was not the same object.

That got me thinking that since we do this query/update in many places, it might be risky to use objects as values. Instead, it will be better if we use the identifier of the object.

I will improve the comment. No need to talk about "bind" actually.

peterpeterparker replied:
Thanks for the feedback. I guess it's because the  HTML `option` value is a `string` (https://developer.mozilla.org/en-US/docs/Web/API/HTMLOptionElement9 - i.e. not support object or maybe I am misunderstanding something.

lmuntaner replied:
Let's do a call on Monday otherwise.

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Dropdown.svelte:
```diff
@@ -0,0 +1,52 @@
+  export let testId: string | undefined = undefined;
+</script>
+
+<div>
```
can't we spare this `div`?

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Dropdown.svelte:
```diff
@@ -0,0 +1,52 @@
+
+<div>
+  <select bind:value {name} data-tid={testId}>
+    {#each options as option}
```
nth: deconstruct option

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Dropdown.svelte:
```diff
@@ -0,0 +1,52 @@
+      padding: var(--padding-2x) var(--padding-3x);
+      border-radius: var(--element-border-radius);
+
+      appearance: none;
```
works in Safari and Firefox?

lmuntaner replied:
good point. I had to change the border because it was set in FF.

I also checked in dark mode, and it wasn't looking so good, so I added a background as well.

More on screenshots

peterpeterparker replied:
thanks

peterpeterparker commented on frontend/svelte/src/lib/components/accounts/SelectAccountDropdown.svelte:
```diff
@@ -15,16 +16,22 @@
 
-  let selectableAccounts: Account[] | undefined;
+  let selectableAccounts: SelectOption[] = [];
+  const convertAccountToSelectOption = (account: Account): SelectOption => ({
```
nth: deconstruct 😉

lmuntaner replied:
hehe, yes

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Dropdown.svelte:
```diff
@@ -0,0 +1,52 @@
+  // Do not allow to bind objects, since they are mutable.
+  export let value: string;
+  export let name: string;
+  export let options: SelectOption[];
```
have you though about passing the options as `slot`? would it make the usage easier?

lmuntaner replied:
Do you mean this:

```
<Dropdown
  name="account"
  bind:value={selectedAccountIdentifier}
  testId="select-account-dropdown"
>
  {#each selectableAccounts as { identifier, name }}
    <option value={identifier}>{name ?? $i18n.accounts.main}</option>
  {/each}
</Dropdown>
```

I like it better now, but if we even want to change the Dropdown to something custom made, then we'll probably need the options as a string of objects. As long as we stay with the native dropdown this is ok.

What do you think?

lmuntaner replied:
Ok, sounds good to me.

lmuntaner replied:
> You like better with slot or without with an array property?

I meant I prefer using `slot`, but I wasn't sure using that directly would enable us to move to a custom dropdown. But if we use our own DropdownItem, it will be possible.

The only part that I don't like is that the DropdownItem is not an independent component and needs to go always inside a Dropdown. But it's worth the coupling.

lmuntaner commented:
I changed the screenshtos.

There is no border and it should have a shadow.

I also changed the background so that it appears different in dark mode.

![Screenshot 2022-07-07 at 08 02 37](https://user-images.githubusercontent.com/4550653/177702799-11ff6f16-0bc6-4270-bca3-62e7e61e198d.jpg)
![Screenshot 2022-07-07 at 08 02 46](https://user-images.githubusercontent.com/4550653/177702802-70ea9b07-9ffb-4c71-93d4-40647cc2ba80.jpg)

lmuntaner commented:
@peterpeterparker please review

peterpeterparker reviewed (approved):
LGTM, thx
